### PR TITLE
[webcamerapl] Add extractor

### PR DESCRIPTION
**Pull request by ghost**
### Description of your *pull request* and other information

</details>

*WebCamera.pl* contains many publicly accessible streams of IP cameras across Poland. The purpose of this extractor is to provide yt-dlp/mpv compatability with this site, for example, to visually assess the weather, smog level, time of day, etc. through a bash command, or to provide a clean stream on a television screen in a local restaurant, or to record/save the stream, etc.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by HobbyistDev** [comment]
Does the video always live ? if that case, you should to set `is_live: True` in `info_dict`

File: yt_dlp/extractor/webcamerapl.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<h1.*?>(.+?)</h1>', webpage, 'title')
+
+        rot13_m3u8_url = self._search_regex(r'data-src="(uggc.+?\.z3h8)"', webpage, 'm3u8 url', fatal=False)
+        if not rot13_m3u8_url:
+            self.raise_no_formats('No video/audio found at the provided url.', expected=True)
+        m3u8_url = decode(rot13_m3u8_url, 'rot-13')
+
+        formats = self._extract_m3u8_formats(m3u8_url, video_id, ext='mp4')
```
[the remark this answers is not in the document]
Comment (milkknife) [reply yt_dlp/extractor/webcamerapl.py]: The streams from this site do not return subtitles. Applied the suggested change anyways.

File: yt_dlp/extractor/webcamerapl.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        return {
+            'id': video_id,
+            'title': title,
+            'formats': formats,
+            'url': m3u8_url,
+            'ext': 'mp4',
```
[the remark this answers is not in the document]
Comment (milkknife) [reply yt_dlp/extractor/webcamerapl.py]: Applied the suggested change. I misunderstood the documentation and was under the impressesion that I was to add as many return values as possible. [CONTRIBUTING.md, step 8](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md) _"Add tests and code for as many as you want."_

**Comment by milkknife** [comment]
> Does the video always live ? if that case, you should to set `is_live: True` in `info_dict`

I tried this, but `is_live` returns `None`.

**Comment by pukkandan** [comment]
> > ``
> 
> I tried this, but `is_live` returns `None`.

wdym? pls elaborate

**Comment by pukkandan** [comment]
I'm unable to download anything, possibly due to some geo-restriction
```
❯ yt-dlp https://gdansk-stare-miasto.webcamera.pl
[Webcamerapl] Extracting URL: https://gdansk-stare-miasto.webcamera.pl
[Webcamerapl] gdansk-stare-miasto: Downloading webpage
[Webcamerapl] gdansk-stare-miasto: Downloading m3u8 information
ERROR: [Webcamerapl] gdansk-stare-miasto: Failed to download m3u8 information: <urlopen error [WinError 10061] No connection could be made because the target machine actively refused it> (caused by URLError(ConnectionRefusedError(10061, 'No connection could be made because the target machine actively refused it', None, 10061, None)))
```

Pls confirm everything's still working on your end after my changes

**Comment by milkknife** [comment]
> > > ``
> > 
> > 
> > I tried this, but `is_live` returns `None`.
> 
> wdym? pls elaborate

I misunderstood what he said as just adding `is_live` to the test `info_dict` and nowhere else.



> I'm unable to download anything, possibly due to some geo-restriction
> 
> ```
> ❯ yt-dlp https://gdansk-stare-miasto.webcamera.pl
> [Webcamerapl] Extracting URL: https://gdansk-stare-miasto.webcamera.pl
> [Webcamerapl] gdansk-stare-miasto: Downloading webpage
> [Webcamerapl] gdansk-stare-miasto: Downloading m3u8 information
> ERROR: [Webcamerapl] gdansk-stare-miasto: Failed to download m3u8 information: <urlopen error [WinError 10061] No connection could be made because the target machine actively refused it> (caused by URLError(ConnectionRefusedError(10061, 'No connection could be made because the target machine actively refused it', None, 10061, None)))
> ```
> 
> Pls confirm everything's still working on your end after my changes

The test fails because of an extra timestamp in the title: `Invalid value for field title, expected 'GDAŃSK - widok na Stare Miasto', got 'GDAŃSK - widok na Stare Miasto 2022-12-08 15:52'` 

otherwise everything still works after your changes 
image

File: yt_dlp/extractor/webcamerapl.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class WebcameraplIE(InfoExtractor):
+    _VALID_URL = r'https?://(?P<id>[\w-]+)\.webcamera\.pl'
+    _TESTS = [{
+        'url': 'https://warszawa-plac-zamkowy.webcamera.pl',
+        'info_dict': {
+            'id': 'warszawa-plac-zamkowy',
+            'ext': 'mp4',
+            'title': 'WIDOK NA PLAC ZAMKOWY W WARSZAWIE',
```
Comment (pukkandan) [inline yt_dlp/extractor/webcamerapl.py:13]:
```suggestion
            'title': r're:WIDOK NA PLAC ZAMKOWY W WARSZAWIE \d{4}-\d{2}-\d{2} \d{2}:\d{2}$',
```
yt-dlp adds timestamp to title for livestreams. Do same for the other test

File: yt_dlp/extractor/webcamerapl.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<h1.*?>(.+?)</h1>', webpage, 'title')
+
+        rot13_m3u8_url = self._search_regex(r'data-src="(uggc.+?\.z3h8)"', webpage, 'm3u8 url', fatal=False)
+        if not rot13_m3u8_url:
+            self.raise_no_formats('No video/audio found at the provided url.', expected=True)
```
[the remark this answers is not in the document]
Comment (milkknife) [reply yt_dlp/extractor/webcamerapl.py]:
```
$ PYTHONPATH=. mpv https://www.webcamera.pl/
[ytdl_hook] ERROR: [Webcamerapl] www: No video/audio found at the provided url
[ytdl_hook] youtube-dl failed: unexpected error occurred
Failed to recognize file format.
```
Any webpage on this site that has no video, no? It is a little more verbose than 'unexpected error' if there were to be a typo, or if a stream were to be removed.

File: yt_dlp/extractor/webcamerapl.py (hunk 3 above)
Comment (milkknife) [reply yt_dlp/extractor/webcamerapl.py]: Sure, added.

**Comment by gamer191** [comment]
Please ignore this comment. It is semi-automated and is intended to ensure PRs by @ghost get indexed by Github

**Pull request merged**